<!DOCTYPE html>
<html>
<head>
<title>main_doc.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<h1 id="secapr---sequence-capture-processor">SeCaPr - Sequence Capture Processor</h1>
<p>A computational pipeline for processing Illumina sequence capture data <a href="http://bioconda.github.io/recipes/secapr/README.html"><img src="https://anaconda.org/bioconda/secapr/badges/downloads.svg" alt="downloads"></a></p>
<hr>
<h2 id="documentation">Documentation</h2>
<p>The documentation, including an empirical data tutorial, is divided into the following steps (click on the links):</p>
<ul>
<li><a href="subdocs/cleaning_trimming.ipynb">Cleaning and adapter trimming</a></li>
<li><a href="subdocs/contig_assembly.ipynb">De-novo contig assembly</a></li>
<li><a href="subdocs/extract_contigs.ipynb">Extract target contigs</a></li>
<li><a href="subdocs/align_contigs.ipynb">Align contigs</a></li>
<li><a href="subdocs/reference_assembly.ipynb">Reference-based assembly</a></li>
<li><a href="subdocs/locus_selection.ipynb">Locus selection</a></li>
<li><a href="subdocs/phasing.ipynb">Phasing alleles</a></li>
<li><a href="subdocs/phylogeny_msc.ipynb">Phylogeny estimation</a></li>
</ul>
<hr>
<h2 id="installation--setup">Installation &amp; Setup</h2>
<p>SECAPR is available as a conda package on the bioconda channel. This makes installation very simple. Follow the instructions on this page to get the SECAPR pipeline set up and ready to use:</p>
<div class="alert alert-block alert-info">
<table>
<thead>
<tr>
<th>INFO: All commands in this tutorial have to be executed from a bash-command line terminal.</th>
</tr>
</thead>
<tbody></tbody>
</table>
</div>
<hr>
<h3 id="1-install-conda">1. Install conda</h3>
<p>Download the <strong>Python2.7 version</strong> of <a href="https://conda.io/miniconda.html">Miniconda</a> and install it by executing the downloaded sh-file (see commands below). Conda is a software and environment manager, that makes installation of new software and of required dependencies very simple and straightforward.</p>
<p><em>Download conda (MacOS 64bit):</em></p>
<div class="alert alert-block alert-info">
wget https://repo.continuum.io/miniconda/Miniconda2-latest-MacOSX-x86_64.sh
</div>
<p><em>Download conda (Linux 64bit):</em></p>
<div class="alert alert-block alert-info">
wget https://repo.continuum.io/miniconda/Miniconda2-latest-Linux-x86_64.sh
</div>
<p><em>Download conda (Linux 32bit):</em></p>
<div class="alert alert-block alert-info">
wget https://repo.continuum.io/miniconda/Miniconda2-latest-Linux-x86.sh
</div>
<p><em>Install conda:</em></p>
<div class="alert alert-block alert-info">
sh Miniconda2-latest-*.sh
</div>
<p><em>Add Bioconda channels (containing bioinformatics software):</em></p>
<div class="alert alert-block alert-info">
conda config --add channels defaults; conda config --add channels conda-forge; conda config --add channels bioconda; conda config --add channels https://conda.anaconda.org/faircloth-lab     
</div>
<hr>
<h3 id="2-install-the-secapr-environment">2. Install the SECAPR environment</h3>
<p>Conda automatically downloads and installs all necessary software dependencies. We strongly recommend to <strong>install SECAPR and all it's dependencies in a separate virtual environment</strong>, in order to not interfer with potentially already installed verisons of the software dependencies.</p>
<p><em>Install SECAPR in virtual environment (here named <code>secapr_env</code>):</em></p>
<div class="alert alert-block alert-info">
<p>conda create -n secapr_env secapr</p>
</div>
<p>Alternatively you can also just plainly install the software on your computer (without creating an environment) by clicking on the icon below and following the instructions (<strong>not recommended!</strong>):</p>
<p><a href="http://bioconda.github.io/recipes/secapr/README.html"><img src="https://img.shields.io/badge/install with-bioconda-brightgreen.svg?style=flat-square" alt="install with bioconda"></a></p>
<hr>
<h3 id="3-activate-the-environment">3. Activate the environment</h3>
<p>To activate the newly created environment, type:</p>
<p><em>Activate environment</em>:</p>
<div class="alert alert-block alert-info">
source activate secapr_env
</div>
<p>When the environment is activated, all the necessary software dependencies will be available in the standarad path, e.g. when you type <code>samtools</code> the samtools version required for SECAPR will be executed. After you are done using secapr, you can deactivate the environment to switch back to your standard environment with this command:</p>
<p><em>De-activate environment</em>:</p>
<div class="alert alert-block alert-info">
<p>source deactivate</p>
</div>
<hr>
<h3 id="4-check-active-environment">4. Check active environment</h3>
<p>Check if you are connected to the correct environment (there should eb a star in front of secapr_env in the output of this command):</p>
<p><em>Check active environment</em>:</p>
<div class="alert alert-block alert-info">
<p>conda info --envs</p>
</div>
<div class="alert alert-block alert-warning">
<p>IMPORTANT : When you are using the SECAPR pipeline, make sure the secapr_env is activated. Activate with <strong>source activate secapr_env</strong></p>
</div>
<hr>
<h3 id="5-install-secapr-development-version">5. Install SECAPR development version</h3>
<p>The development version of SECAPR is stored on this GitHub page and contains the newest updates, which might not yet be available through the conda version. However you need to install the SECAPR environment with conda first by following the steps above. Once the environment is installed, you can update SECAPR to the development version by following these steps:</p>
<ol>
<li>Connect to your secapr environment (<code>source activate secapr_env</code>)</li>
<li>Remove the current secapr installation (<code>conda remove secapr</code>)</li>
<li>Download the new version from github (<code>wget https://github.com/AntonelliLab/seqcap_processor/archive/master.zip</code>)</li>
<li>Unzip the downloaded file (<code>unzip master.zip</code>)</li>
<li>Move the unzipped directory to a safe location on your computer, i.e. not on your Desktop or Download folder. This will be the path where secapr will be executed from in the future</li>
<li>Enter the unzipped secapr directory (<code>cd seqcap_processor-master</code>)</li>
<li>Install secapr from the folder (<code>python -m pip install -e .</code>)</li>
</ol>

</body>
</html>
