<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Welcome to seqcap_processor’s documentation! &mdash; seqcap_processor 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="seqcap_processor 1.0 documentation" href="#" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

  </head>
  <body role="document">  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="welcome-to-seqcap-processor-s-documentation">
<h1>Welcome to seqcap_processor&#8217;s documentation!<a class="headerlink" href="#welcome-to-seqcap-processor-s-documentation" title="Permalink to this headline">¶</a></h1>
</div>
<div class="section" id="phasing-uce-data">
<h1>Phasing UCE data<a class="headerlink" href="#phasing-uce-data" title="Permalink to this headline">¶</a></h1>
<p>In the previous step we extracted the target FASTA sequences from the assembly files, which represent UCE loci. The resulting FASTA file (e.g. <code class="docutils literal"><span class="pre">dataset1.fasta</span></code>) contains all contig sequences of interest. However, any diploid individual may have more than one sequence at each individual UCE locus. This allelic information lies in the reads for each sample and can be recovered by allele phasing.</p>
<p>In this approach we use the fastq read data and map them against the recovered UCE contig sequences. This collects all the read variation at each given locus for each individual. In a subsequent step all matching reads for each locus are phased into two separate allele bins. We then create a consensus sequence of each allele bin separately and export all allele FASTA sequences into a joined FASTA library.</p>
<div class="section" id="separating-the-contig-fasta-library">
<h2>Separating the contig FASTA library<a class="headerlink" href="#separating-the-contig-fasta-library" title="Permalink to this headline">¶</a></h2>
<p>First we will have to split the contig FASTA file (e.g. <code class="docutils literal"><span class="pre">dataset1.fasta</span></code>) into sample specific FASTA databases, which will serve as the sample specific templates for the following mapping. You can do this in two simple steps:</p>
<ol class="arabic simple">
<li>Deposit all sample names/IDs into a text file named <code class="docutils literal"><span class="pre">sample_IDs.txt</span></code>. Each sample ID should occupy an individual line in the text file. This file will be used in the next step by iterating through each line and extracting all sequences containing the stated sample ID in the FASTA header inot a separate FASTA file. Make sure to use the exact sample IDs as they occur in the contig FASTA file.</li>
</ol>
<div class="admonition attention">
<p class="first admonition-title">Attention</p>
<p class="last">Make sure that the sample IDs are unique. This can be tricky when having names like <code class="docutils literal"><span class="pre">genus_species1</span></code> and <code class="docutils literal"><span class="pre">genus_species10</span></code>, since when searching for the former one (<code class="docutils literal"><span class="pre">genus_species1</span></code>) all sequences for both samples (<code class="docutils literal"><span class="pre">genus_species1</span></code> and <code class="docutils literal"><span class="pre">genus_species10</span></code>) will be returned. To avoid this it is recommendable to add an underscore after each sample ID (e.g. <code class="docutils literal"><span class="pre">genus_species1_</span></code>).</p>
</div>
<ol class="arabic simple" start="2">
<li>Run the following command in the command line after altering the paths to the contig file (<code class="docutils literal"><span class="pre">dataset1.fasta</span></code>) and to the output fasta files (<code class="docutils literal"><span class="pre">${sample}_contigs.fasta</span></code>):</li>
</ol>
<div class="highlight-bash"><div class="highlight"><pre><span class="k">for</span> sample in <span class="k">$(</span>cat sample_IDs.txt<span class="k">)</span><span class="p">;</span> <span class="se">\</span>
    <span class="k">do</span> grep <span class="nv">$sample</span> -A <span class="m">1</span> /path/to/uce/taxon-set1/dataset1.fasta &gt; /path/to/uce/taxon-set1/sample_specific/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span>_contigs.fasta<span class="p">;</span> <span class="se">\</span>
    <span class="k">done</span>
</pre></div>
</div>
</div>
<div class="section" id="creating-a-configuration-file">
<h2>Creating a configuration file<a class="headerlink" href="#creating-a-configuration-file" title="Permalink to this headline">¶</a></h2>
<p>Before you run the script, you have to create a configuration file, telling the program where the cleaned and trimmed fastq reads are stored for each sample and where to find the contig FASTA library for each sample.
The configuration file should look like in the following example and should be saved as e.g. <code class="docutils literal"><span class="pre">phasing.conf</span></code>:</p>
<div class="highlight-python"><div class="highlight"><pre>[references]
genus_species1:/path/to/uce/taxon-set1/sample_specific/genus_species1_contigs.fasta
genus_species2:/path/to/uce/taxon-set1/sample_specific/genus_species2_contigs.fasta

[individuals]
genus_species1:/path/to/clean-fastq/genus_species1
genus_species2:/path/to/clean-fastq/genus_species2

[flowcell]
genus_species1:XXYYZZ
genus_species2:XXYYZZ
</pre></div>
</div>
<div class="section" id="references">
<h3>[references]<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<p>In this section you simply state the sample ID (<code class="docutils literal"><span class="pre">genus_species1</span></code>) followed by a colon (<code class="docutils literal"><span class="pre">:</span></code>) and the full path to the sample-specific FASTA library which was generated in the previous step.</p>
</div>
<div class="section" id="individuals">
<h3>[individuals]<a class="headerlink" href="#individuals" title="Permalink to this headline">¶</a></h3>
<p>In this section you give the complete path to the cleaned and trimmed reads folder for each sample.</p>
<div class="admonition attention">
<p class="first admonition-title">Attention</p>
<p class="last">The cleaned reads used by this program should preferably be generated by <a class="reference external" href="https://github.com/faircloth-lab/illumiprocessor/">illumiprocessor</a> as the folder structure of the cleaned reads files is assumed to be that of <a class="reference external" href="https://github.com/faircloth-lab/illumiprocessor/">illumiprocessor</a> . This means that the zipped fastq files (fastq.gz) have to be located in a subfolder with the name <code class="docutils literal"><span class="pre">split-adapter-quality-trimmed</span></code> within each sample-specific folder.</p>
</div>
</div>
<div class="section" id="flowcell">
<h3>[flowcell]<a class="headerlink" href="#flowcell" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="mapping-reads-against-contigs">
<h2>Mapping reads against contigs<a class="headerlink" href="#mapping-reads-against-contigs" title="Permalink to this headline">¶</a></h2>
<p>To map the fastq read files against the contig reference database for each sample, run:</p>
<div class="highlight-bash"><div class="highlight"><pre>phyluce_snp_bwa_multiple_align <span class="se">\</span>
    --config /path/to/phasing.conf <span class="se">\</span>
    --output /path/to/mapping_results <span class="se">\</span>
    --subfolder split-adapter-quality-trimmed
</pre></div>
</div>
<p>This will produce an output along these lines:</p>
<div class="highlight-python"><div class="highlight"><pre>2016-03-09 16:40:22,628 - phyluce_snp_bwa_multiple_align - INFO - ============ Starting phyluce_snp_bwa_multiple_align ============
2016-03-09 16:40:22,628 - phyluce_snp_bwa_multiple_align - INFO - Version: 1.5.0
2016-03-09 16:40:22,629 - phyluce_snp_bwa_multiple_align - INFO - Argument --config: /path/to/phasing.conf
2016-03-09 16:40:22,629 - phyluce_snp_bwa_multiple_align - INFO - Argument --cores: 1
2016-03-09 16:40:22,629 - phyluce_snp_bwa_multiple_align - INFO - Argument --log_path: None
2016-03-09 16:40:22,629 - phyluce_snp_bwa_multiple_align - INFO - Argument --mem: False
2016-03-09 16:40:22,629 - phyluce_snp_bwa_multiple_align - INFO - Argument --no_remove_duplicates: False
2016-03-09 16:40:22,629 - phyluce_snp_bwa_multiple_align - INFO - Argument --output: /path/to/mapping_results
2016-03-09 16:40:22,629 - phyluce_snp_bwa_multiple_align - INFO - Argument --subfolder: split-adapter-quality-trimmed
2016-03-09 16:40:22,629 - phyluce_snp_bwa_multiple_align - INFO - Argument --verbosity: INFO
2016-03-09 16:40:22,630 - phyluce_snp_bwa_multiple_align - INFO - ============ Starting phyluce_snp_bwa_multiple_align ============
2016-03-09 16:40:22,631 - phyluce_snp_bwa_multiple_align - INFO - Getting input filenames and creating output directories
2016-03-09 16:40:22,633 - phyluce_snp_bwa_multiple_align - INFO - ---------------------- Processing genus_species1 ----------------------
2016-03-09 16:40:22,633 - phyluce_snp_bwa_multiple_align - INFO - Finding fastq/fasta files
2016-03-09 16:40:22,636 - phyluce_snp_bwa_multiple_align - INFO - File type is fastq
2016-03-09 16:40:22,637 - phyluce_snp_bwa_multiple_align - INFO - Creating read index file for genus_species1-READ1.fastq.gz
2016-03-09 16:40:33,999 - phyluce_snp_bwa_multiple_align - INFO - Creating read index file for genus_species1-READ2.fastq.gz
2016-03-09 16:40:45,142 - phyluce_snp_bwa_multiple_align - INFO - Building BAM for genus_species1
2016-03-09 16:41:33,195 - phyluce_snp_bwa_multiple_align - INFO - Cleaning BAM for genus_species1
2016-03-09 16:42:03,410 - phyluce_snp_bwa_multiple_align - INFO - Adding RG header to BAM for genus_species1
2016-03-09 16:42:49,518 - phyluce_snp_bwa_multiple_align - INFO - Marking read duplicates from BAM for genus_species1
2016-03-09 16:43:26,917 - phyluce_snp_bwa_multiple_align - INFO - Creating read index file for genus_species1-READ-singleton.fastq.gz
2016-03-09 16:43:27,066 - phyluce_snp_bwa_multiple_align - INFO - Building BAM for genus_species1
2016-03-09 16:43:27,293 - phyluce_snp_bwa_multiple_align - INFO - Cleaning BAM for genus_species1
2016-03-09 16:43:27,748 - phyluce_snp_bwa_multiple_align - INFO - Adding RG header to BAM for genus_species1
2016-03-09 16:43:28,390 - phyluce_snp_bwa_multiple_align - INFO - Marking read duplicates from BAM for genus_species1
2016-03-09 16:43:30,633 - phyluce_snp_bwa_multiple_align - INFO - Merging BAMs for genus_species1
2016-03-09 16:44:05,811 - phyluce_snp_bwa_multiple_align - INFO - Indexing BAM for genus_species1
2016-03-09 16:44:08,047 - phyluce_snp_bwa_multiple_align - INFO - ---------------------- Processing genus_species2 ----------------------
...
</pre></div>
</div>
</div>
<div class="section" id="phasing-mapped-reads">
<h2>Phasing mapped reads<a class="headerlink" href="#phasing-mapped-reads" title="Permalink to this headline">¶</a></h2>
<p>In the previous step you mapped the reads against the contig FASTA file for each sample. The results are stored in the output folder in bam-format. Now you can start the actual phasing of the reads. This will sort the reads within each bam file into two separate bam files (<code class="docutils literal"><span class="pre">genus_species1.0.bam</span></code> and <code class="docutils literal"><span class="pre">genus_species1.1.bam</span></code>).
The program is very easy to run and just requires the path to the bam files (output folder from previous mapping program, <code class="docutils literal"><span class="pre">/path/to/mapping_results</span></code>) and the path to the configuration file, which is the same file as used in the previous step (<code class="docutils literal"><span class="pre">/path/to/phasing.conf</span></code>):</p>
<div class="highlight-bash"><div class="highlight"><pre>phyluce_snp_phase_uces <span class="se">\</span>
    --config /path/to/phasing.conf <span class="se">\</span>
    --bams /path/to/mapping_results/ <span class="se">\</span>
    --output /path/to/phased_reads
</pre></div>
</div>
<p>The output is supposed to look like this:</p>
<div class="highlight-python"><div class="highlight"><pre>2016-03-09 17:31:43,790 - phyluce_snp_phase_uces - INFO - ================ Starting phyluce_snp_phase_uces ================
2016-03-09 17:31:43,790 - phyluce_snp_phase_uces - INFO - Version: 1.5.0
2016-03-09 17:31:43,790 - phyluce_snp_phase_uces - INFO - Argument --bams: /path/to/mapping_results/
2016-03-09 17:31:43,790 - phyluce_snp_phase_uces - INFO - Argument --config: /path/to/phasing.conf
2016-03-09 17:31:43,791 - phyluce_snp_phase_uces - INFO - Argument --conservative: False
2016-03-09 17:31:43,791 - phyluce_snp_phase_uces - INFO - Argument --cores: 1
2016-03-09 17:31:43,791 - phyluce_snp_phase_uces - INFO - Argument --log_path: None
2016-03-09 17:31:43,791 - phyluce_snp_phase_uces - INFO - Argument --output: /path/to/phased_reads
2016-03-09 17:31:43,791 - phyluce_snp_phase_uces - INFO - Argument --verbosity: INFO
2016-03-09 17:31:43,791 - phyluce_snp_phase_uces - INFO - ================ Starting phyluce_snp_phase_uces ================
2016-03-09 17:31:43,793 - phyluce_snp_phase_uces - INFO - Getting input filenames and creating output directories
2016-03-09 17:41:32,196 - phyluce_snp_phase_uces - INFO - ----------------------- Processing genus_species1 ----------------------
2016-03-09 17:41:32,196 - phyluce_snp_phase_uces - INFO - Phasing BAM file for genus_species1
2016-03-09 17:41:42,787 - phyluce_snp_phase_uces - INFO - Sorting BAM for genus_species1
2016-03-09 17:41:44,239 - phyluce_snp_phase_uces - INFO - Sorting BAM for genus_species1
2016-03-09 17:41:45,705 - phyluce_snp_phase_uces - INFO - Creating REF/ALT allele FASTQ file 0
2016-03-09 17:42:02,203 - phyluce_snp_phase_uces - INFO - Creating REF/ALT allele FASTQ file 1
2016-03-09 17:42:18,776 - phyluce_snp_phase_uces - INFO - Creating REF/ALT allele FASTQ file unphased
2016-03-09 17:42:58,258 - phyluce_snp_phase_uces - INFO - Creating REF/ALT allele FASTA file 0 from FASTQ 0
2016-03-09 17:42:58,273 - phyluce_snp_phase_uces - INFO - Creating REF/ALT allele FASTA file 1 from FASTQ 1
2016-03-09 17:42:58,286 - phyluce_snp_phase_uces - INFO - Creating REF/ALT allele FASTA file unphased from FASTQ unphased
2016-03-09 17:42:58,298 - phyluce_snp_phase_uces - INFO - Checking for correct FASTA files
2016-03-09 17:42:58,298 - phyluce_snp_phase_uces - INFO - Cleaning FASTA files
2016-03-09 17:42:58,475 - phyluce_snp_phase_uces - INFO - Balancing FASTA files
2016-03-09 17:42:58,627 - phyluce_snp_phase_uces - INFO - Symlinking FASTA files
2016-03-09 17:42:58,627 - phyluce_snp_phase_uces - INFO - ---------------------- Processing genus_species2 ---------------------
2016-03-09 17:42:58,628 - phyluce_snp_phase_uces - INFO - Phasing BAM file for genus_species2
2016-03-09 17:43:02,459 - phyluce_snp_phase_uces - INFO - Sorting BAM for genus_species2
2016-03-09 17:43:03,012 - phyluce_snp_phase_uces - INFO - Sorting BAM for genus_species2
2016-03-09 17:43:03,565 - phyluce_snp_phase_uces - INFO - Creating REF/ALT allele FASTQ file 0
2016-03-09 17:43:11,131 - phyluce_snp_phase_uces - INFO - Creating REF/ALT allele FASTQ file 1
2016-03-09 17:43:18,723 - phyluce_snp_phase_uces - INFO - Creating REF/ALT allele FASTQ file unphased
2016-03-09 17:43:37,441 - phyluce_snp_phase_uces - INFO - Creating REF/ALT allele FASTA file 0 from FASTQ 0
2016-03-09 17:43:37,454 - phyluce_snp_phase_uces - INFO - Creating REF/ALT allele FASTA file 1 from FASTQ 1
2016-03-09 17:43:37,464 - phyluce_snp_phase_uces - INFO - Creating REF/ALT allele FASTA file unphased from FASTQ unphased
2016-03-09 17:43:37,472 - phyluce_snp_phase_uces - INFO - Checking for correct FASTA files
2016-03-09 17:43:37,473 - phyluce_snp_phase_uces - INFO - Cleaning FASTA files
2016-03-09 17:43:37,633 - phyluce_snp_phase_uces - INFO - Balancing FASTA files
2016-03-09 17:43:37,776 - phyluce_snp_phase_uces - INFO - Symlinking FASTA files
2016-03-09 17:43:37,779 - phyluce_snp_phase_uces - INFO - ------------------ Merging alleles from all loci-----------------
2016-03-09 17:43:38,577 - phyluce_snp_phase_uces - INFO - Wrote 819 loci for genus_species1
2016-03-09 17:43:38,669 - phyluce_snp_phase_uces - INFO - Wrote 812 loci for genus_species2
2016-03-09 17:43:38,675 - phyluce_snp_phase_uces - INFO - ================ Completed phyluce_snp_phase_uces ===============
</pre></div>
</div>
<p>The program automatically produces a consensus sequence for each of these phased bam files (= allele sequence) and stores these allele sequences of all samples in a joined FASTA file (<code class="docutils literal"><span class="pre">joined_allele_sequences_all_samples.fasta</span></code>). This allele FASTA is deposited in the subfolder <code class="docutils literal"><span class="pre">fastas</span></code> within your output folder (e.g. <code class="docutils literal"><span class="pre">/path/to/phased_reads</span></code>) and can be used as input for the following alignment steps.</p>
</div>
</div>
<div class="section" id="aligning-and-trimming-fasta-data">
<h1>Aligning and trimming FASTA data<a class="headerlink" href="#aligning-and-trimming-fasta-data" title="Permalink to this headline">¶</a></h1>
<p>With all of that out of the way, things get much easier to deal with.  Now, we
need to align our data across loci, and once we&#8217;re done with that, the remaining
operations we can run on the data are format-conversions, QC steps, matrix
trimming for completeness, and any number of other fun things.</p>
<p>Aligning the amount of data generated by enrichment approaches is reasonably
computationally intensive - so the alignment step goes fastest if you have a
multicore machine.  You also have several alignment options available, although
I would suggest sticking with MAFFT.</p>
<div class="admonition attention">
<p class="first admonition-title">Attention</p>
<p>The alignment process, as implemented by <a href="#id1"><span class="problematic" id="id2">phyluce_</span></a>, includes
trimming steps that trim ragged edges and remove alignments that become to
short following trimming.</p>
<p class="last">To turn trimming off and trim alignments using another approach, pass the
<code class="docutils literal"><span class="pre">--no-trim</span></code> option.  There are also several more options related to
trimming that you can tweak.  To view these, run
<code class="docutils literal"><span class="pre">phyluce_align_seqcap_align</span> <span class="pre">--help</span></code>.</p>
</div>
<div class="section" id="complete-data-matrix">
<h2>Complete data matrix<a class="headerlink" href="#complete-data-matrix" title="Permalink to this headline">¶</a></h2>
<div class="section" id="alignment">
<h3>Alignment<a class="headerlink" href="#alignment" title="Permalink to this headline">¶</a></h3>
<p>To align the loci, by taxon, in the FASTA file you just created, run:</p>
<div class="highlight-bash"><div class="highlight"><pre>phyluce_align_seqcap_align <span class="se">\</span>
    --fasta /path/to/uce/taxon-set1/dataset1.fasta <span class="se">\</span>
    --output /path/to/uce/taxon-set1/mafft-nexus/ <span class="se">\</span>
    --taxa <span class="m">3</span> <span class="se">\</span>
    --aligner mafft <span class="se">\</span>
    --cores 8
</pre></div>
</div>
<div class="admonition attention">
<p class="first admonition-title">Attention</p>
<p class="last">If you pass more <code class="docutils literal"><span class="pre">--cores</span></code> than your machine has, you will
receive an error.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Here, we are accepting the default, output alignment format (&#8220;nexus&#8221;).
To change that format to something else, pass the <code class="docutils literal"><span class="pre">--output-format</span></code> option
with a choice of {fasta,nexus,phylip,clustal,emboss,stockholm}.</p>
</div>
</div>
<div class="section" id="alignment-stats">
<h3>Alignment stats<a class="headerlink" href="#alignment-stats" title="Permalink to this headline">¶</a></h3>
<p>Once you have alignments, it&#8217;s nice to get a general sense of their length and
composition.  You can quickly (with a multicore machine) summarize thousands of
alignments by running:</p>
<div class="highlight-bash"><div class="highlight"><pre>phyluce_align_get_align_summary_data <span class="se">\</span>
    --alignments /path/to/uce/taxon-set1/mafft-nexus/ <span class="se">\</span>
    --cores 12
</pre></div>
</div>
<p>This will produce output that looks similar to:</p>
<div class="highlight-python"><div class="highlight"><pre>2014-04-24 17:31:15,724 - get_align_summary_data - INFO - ================ Starting get_align_summary_data ================
2014-04-24 17:31:15,724 - get_align_summary_data - INFO - Version: git 7aec8f1
2014-04-24 17:31:15,724 - get_align_summary_data - INFO - Argument --alignments: /path/to/uce/taxon-set1/mafft-nexus/
2014-04-24 17:31:15,724 - get_align_summary_data - INFO - Argument --cores: 12
2014-04-24 17:31:15,724 - get_align_summary_data - INFO - Argument --input_format: nexus
2014-04-24 17:31:15,724 - get_align_summary_data - INFO - Argument --log_path: /path/to/uce/taxon-set1/log
2014-04-24 17:31:15,725 - get_align_summary_data - INFO - Argument --show_taxon_counts: False
2014-04-24 17:31:15,725 - get_align_summary_data - INFO - Argument --verbosity: INFO
2014-04-24 17:31:15,725 - get_align_summary_data - INFO - Getting alignment files
2014-04-24 17:31:15,729 - get_align_summary_data - INFO - Computing summary statistics using 12 cores
2014-04-24 17:31:16,653 - get_align_summary_data - INFO - ----------------------- Alignment summary -----------------------
2014-04-24 17:31:16,654 - get_align_summary_data - INFO - [Alignments] loci:    306
2014-04-24 17:31:16,654 - get_align_summary_data - INFO - [Alignments] length:  223,929
2014-04-24 17:31:16,654 - get_align_summary_data - INFO - [Alignments] mean:    731.79
2014-04-24 17:31:16,654 - get_align_summary_data - INFO - [Alignments] 95% CI:  17.01
2014-04-24 17:31:16,654 - get_align_summary_data - INFO - [Alignments] min:     275
2014-04-24 17:31:16,654 - get_align_summary_data - INFO - [Alignments] max:     1,109
2014-04-24 17:31:16,655 - get_align_summary_data - INFO - ------------------------- Taxon summary -------------------------
2014-04-24 17:31:16,655 - get_align_summary_data - INFO - [Taxa] mean:          27.00
2014-04-24 17:31:16,655 - get_align_summary_data - INFO - [Taxa] 95% CI:        0.00
2014-04-24 17:31:16,656 - get_align_summary_data - INFO - [Taxa] min:           27
2014-04-24 17:31:16,656 - get_align_summary_data - INFO - [Taxa] max:           27
2014-04-24 17:31:16,656 - get_align_summary_data - INFO - ----------------- Missing data from trim summary ----------------
2014-04-24 17:31:16,656 - get_align_summary_data - INFO - [Missing] mean:       7.61
2014-04-24 17:31:16,656 - get_align_summary_data - INFO - [Missing] 95% CI:     0.24
2014-04-24 17:31:16,656 - get_align_summary_data - INFO - [Missing] min:        1.13
2014-04-24 17:31:16,657 - get_align_summary_data - INFO - [Missing] max:        15.79
2014-04-24 17:31:16,661 - get_align_summary_data - INFO - -------------------- Character count summary --------------------
2014-04-24 17:31:16,661 - get_align_summary_data - INFO - [All characters]      6,046,083
2014-04-24 17:31:16,661 - get_align_summary_data - INFO - [Nucleotides]         4,924,129
2014-04-24 17:31:16,661 - get_align_summary_data - INFO - ---------------- Data matrix completeness summary ---------------
2014-04-24 17:31:16,661 - get_align_summary_data - INFO - [Matrix 50%]          306 alignments
2014-04-24 17:31:16,661 - get_align_summary_data - INFO - [Matrix 55%]          306 alignments
2014-04-24 17:31:16,662 - get_align_summary_data - INFO - [Matrix 60%]          306 alignments
2014-04-24 17:31:16,662 - get_align_summary_data - INFO - [Matrix 65%]          306 alignments
2014-04-24 17:31:16,662 - get_align_summary_data - INFO - [Matrix 70%]          306 alignments
2014-04-24 17:31:16,662 - get_align_summary_data - INFO - [Matrix 75%]          306 alignments
2014-04-24 17:31:16,662 - get_align_summary_data - INFO - [Matrix 80%]          306 alignments
2014-04-24 17:31:16,662 - get_align_summary_data - INFO - [Matrix 85%]          306 alignments
2014-04-24 17:31:16,662 - get_align_summary_data - INFO - [Matrix 90%]          306 alignments
2014-04-24 17:31:16,662 - get_align_summary_data - INFO - [Matrix 95%]          306 alignments
2014-04-24 17:31:16,663 - get_align_summary_data - INFO - ------------------------ Character counts -----------------------
2014-04-24 17:31:16,663 - get_align_summary_data - INFO - [Characters] &#39;-&#39; is present 651,009 times
2014-04-24 17:31:16,663 - get_align_summary_data - INFO - [Characters] &#39;?&#39; is present 470,945 times
2014-04-24 17:31:16,663 - get_align_summary_data - INFO - [Characters] &#39;A&#39; is present 1,386,821 times
2014-04-24 17:31:16,663 - get_align_summary_data - INFO - [Characters] &#39;C&#39; is present 1,089,729 times
2014-04-24 17:31:16,663 - get_align_summary_data - INFO - [Characters] &#39;G&#39; is present 1,094,159 times
2014-04-24 17:31:16,663 - get_align_summary_data - INFO - [Characters] &#39;T&#39; is present 1,353,420 times
2014-04-24 17:31:16,664 - get_align_summary_data - INFO - ================ Completed get_align_summary_data ===============
</pre></div>
</div>
</div>
<div class="section" id="locus-name-removal">
<h3>Locus name removal<a class="headerlink" href="#locus-name-removal" title="Permalink to this headline">¶</a></h3>
<p>For historical reasons, and also for users to ensure that the sequence data
aligned together are from the same loci, each sequence line in the alignment
file output by <code class="docutils literal"><span class="pre">seqcap_align_2</span></code> contains the <code class="docutils literal"><span class="pre">genus_species1</span></code> designator,
but the <code class="docutils literal"><span class="pre">genus_species1</span></code> designator is also prepended with the locus name
(e.g. <code class="docutils literal"><span class="pre">uce-1005_genus_species1</span></code>).  We need to remove these if we plan to
concatenate the loci (<span class="xref std std-ref">raxml-concat</span>).  More generally, it is a good idea
to remove locus names from sequence lines before running any analyses. To do
this, run:</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>phyluce_align_remove_locus_name_from_nexus_lines <span class="se">\</span>
    --alignments /path/to/uce/taxon-set1/mafft-nexus/ <span class="se">\</span>
    --output /path/to/uce/taxon-set1/mafft-nexus-clean/ <span class="se">\</span>
    --taxa 3
</pre></div>
</div>
</div></blockquote>
<p>Overview:</p>
<div class="toctree-wrapper compound">
<ul class="simple">
</ul>
</div>
</div>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span>Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span>Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span>Search Page</span></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Welcome to seqcap_processor&#8217;s documentation!</a></li>
<li><a class="reference internal" href="#phasing-uce-data">Phasing UCE data</a><ul>
<li><a class="reference internal" href="#separating-the-contig-fasta-library">Separating the contig FASTA library</a></li>
<li><a class="reference internal" href="#creating-a-configuration-file">Creating a configuration file</a><ul>
<li><a class="reference internal" href="#references">[references]</a></li>
<li><a class="reference internal" href="#individuals">[individuals]</a></li>
<li><a class="reference internal" href="#flowcell">[flowcell]</a></li>
</ul>
</li>
<li><a class="reference internal" href="#mapping-reads-against-contigs">Mapping reads against contigs</a></li>
<li><a class="reference internal" href="#phasing-mapped-reads">Phasing mapped reads</a></li>
</ul>
</li>
<li><a class="reference internal" href="#aligning-and-trimming-fasta-data">Aligning and trimming FASTA data</a><ul>
<li><a class="reference internal" href="#complete-data-matrix">Complete data matrix</a><ul>
<li><a class="reference internal" href="#alignment">Alignment</a></li>
<li><a class="reference internal" href="#alignment-stats">Alignment stats</a></li>
<li><a class="reference internal" href="#locus-name-removal">Locus name removal</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2016, Tobias Hofmann.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.6</a>
      
      |
      <a href="_sources/index.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>